source/extensions/http/header_validators/envoy_default/http1_header_validator.cc

```diff
@@ -0,0 +1,368 @@
+#include "source/extensions/http/header_validators/envoy_default/http1_header_validator.h"
+
+#include "source/extensions/http/header_validators/envoy_default/character_tables.h"
+
+#include "absl/container/node_hash_set.h"
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Http {
+namespace HeaderValidators {
+namespace EnvoyDefault {
+
+using ::envoy::extensions::http::header_validators::envoy_default::v3::HeaderValidatorConfig;
+using ::Envoy::Http::HeaderString;
+using ::Envoy::Http::Protocol;
+using ::Envoy::Http::RequestHeaderMap;
+using HeaderValidatorFunction = ::Envoy::Http::HeaderValidator::HeaderEntryValidationResult (
+    Http1HeaderValidator::*)(const HeaderString&);
+
+struct Http1ResponseCodeDetailValues {
+  const std::string InvalidTransferEncoding = "uhv.http1.invalid_transfer_encoding";
+  const std::string TransferEncodingNotAllowed = "uhv.http1.transfer_encoding_not_allowed";
+  const std::string ContentLengthNotAllowed = "uhv.http1.content_length_not_allowed";
+  const std::string ChunkedContentLength = "uhv.http1.content_length_and_chunked_not_allowed";
+};
+
+using Http1ResponseCodeDetail = ConstSingleton<Http1ResponseCodeDetailValues>;
+
+/*
+ * Header validation implementation for the Http/1 codec. This class follows guidance from
+ * several RFCS:
+ *
+ * RFC 3986 <https://datatracker.ietf.org/doc/html/rfc3986> URI Generic Syntax
+ * RFC 7230 <https://datatracker.ietf.org/doc/html/rfc7230> HTTP/1.1 Message Syntax
+ * RFC 7231 <https://datatracker.ietf.org/doc/html/rfc7231> HTTP/1.1 Semantics and Content
+ *
+ */
+Http1HeaderValidator::Http1HeaderValidator(const HeaderValidatorConfig& config, Protocol protocol,
+                                           StreamInfo::StreamInfo& stream_info)
+    : HeaderValidator(config, protocol, stream_info) {}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+Http1HeaderValidator::validateRequestHeaderEntry(const HeaderString& key,
+                                                 const HeaderString& value) {
+  static const absl::node_hash_map<absl::string_view, HeaderValidatorFunction> kHeaderValidatorMap{
+      {":method", &Http1HeaderValidator::validateMethodHeader},
+      {":authority", &Http1HeaderValidator::validateHostHeader},
+      {"host", &Http1HeaderValidator::validateHostHeader},
```

Comment:
wait, we allow both :authority and host headers?  that seems wrong

Reply:
Ah yes, good catch. "host" is mapped to ":authority" [here](https://github.com/envoyproxy/envoy/blob/eaa674d0c1eadefe872ab89402e28eaeb1d279d4/source/common/http/headers.h#L203) so both keys point to the same value. You can just remove the "host" entry.